### PR TITLE
[Sc 95269] Improve linear superposition handling of built-in CircuitComponents

### PR DESCRIPTION
**Pull request by apchytr**
**Context:**
To be merged after #626.

Our current approach for handling linear superpositions of built in CircuitComponents is not ideal and does not work with optimizations. In this PR, we update `CircuitComponent.__add__` to handle the case appropriately thus enabling us to optimize e.g. cat states. Specifically, what happens now is the parameters of the built-in are concatenated and a new component of the same type is initialized with the `_lin_sup` flag set to `True`. This should also be more efficient as we avoid generating the triple and (possibly) padding things. 

**Description of the Change:** Updated `CircuitComponent.__add__` to handle the case of built-in CircuitComponents. `Dgate.fock_array` and `BSgate.fock_array` are fixed to work with lin sup. Removed `to_fock` on `Dgate` and `BSgate` as they were redundant. `PolyExpAnsatz._fn_kwargs` is renamed to the same attribute as the `Ansatz` i.e. `_kwargs`. Reordered `ParameterSet`. Tests.

**Benefits:** More efficient approach to linear superposition for built-ins that supports optimizations. Some cleanup.

**Possible Drawbacks:** `Number` is still not supported as it isn't batchable (see sc-95331). `normalize` does not work with the optimizations (see sc-95330)

### CODE REVIEW

File: mrmustard/lab/circuit_components.py [hunk 1]
```diff
@@ -805,6 +798,34 @@ def __add__(self, other: CircuitComponent) -> CircuitComponent:
         """
         if self.wires != other.wires:
             raise ValueError("Cannot add components with different wires.")
+
+        # TODO: [sc-95331]
+        if (
```
Comment (apchytr) [inline mrmustard/lab/circuit_components.py:803]: Add check for batched params

File: mrmustard/lab/circuit_components.py [hunk 2]
```diff
@@ -797,6 +798,33 @@ def __add__(self, other: CircuitComponent) -> CircuitComponent:
...
+            isinstance(self.ansatz, PolyExpAnsatz)
+            and self.ansatz._fn is not None
+            and self.ansatz._fn == other.ansatz._fn
+        ):
+            new_params = {}
+            for name in self.ansatz._kwargs:
+                self_param = getattr(self.parameters, name)
+                other_param = getattr(other.parameters, name)
+                if type(self_param) is not type(other_param):
```
Comment (heltluke) [inline mrmustard/lab/circuit_components.py:811]: Should `isinstance` be used here?

File: mrmustard/lab/transformations/bsgate.py [hunk 3]
```diff
@@ -156,16 +157,3 @@ def fock_array(
                 method=method,
             )
         return ret
-
-    def to_fock(self, shape: int | Sequence[int] | None = None) -> BSgate:
```
Comment (heltluke) [inline mrmustard/lab/transformations/bsgate.py:160]: Missed in some other cleanup?

File: tests/test_lab/test_circuit_components.py [hunk 4]
```diff
@@ -219,12 +219,41 @@ def test_to_fock_bargmann_poly_exp(self):
         assert math.allclose(barg_cc.ansatz.c, fock_cc.ansatz.data)
 
     def test_add(self):
+        cc1 = CircuitComponent.from_bargmann(Abc_triple(1), modes_out_ket=(0,))
+        cc2 = CircuitComponent.from_bargmann(Abc_triple(1), modes_out_ket=(0,))
+
+        cc12 = cc1 + cc2
+
```
Comment (heltluke) [inline tests/test_lab/test_circuit_components.py:226]:
```suggestion

        assert isinstance(cc12, CircuitComponent)
```
?

File: tests/test_lab/test_circuit_components.py [hunk 5]
```diff
@@ -219,12 +219,41 @@ def test_to_fock_bargmann_poly_exp(self):
...
+        assert d12.ansatz._lin_sup is True
         assert d12.ansatz == d1.ansatz + d2.ansatz
 
+    def test_add_error(self):
+        d1 = Dgate(1, x=0.1, y=0.1)
+        d2 = Dgate(2, x=0.2, y=0.2)
+        d3 = Dgate(1, x=0.1, y=0.1, x_trainable=True)
+        d4 = Dgate(1, x=0.1, y=0.1, x_trainable=True, x_bounds=(0, 1))
+
+        with pytest.raises(ValueError):
```
Comment (heltluke) [inline tests/test_lab/test_circuit_components.py:248]: Is there no error text worth matching for here? Perhaps we could provide some?

File: tests/test_lab/test_transformations/test_bsgate.py [hunk 6]
```diff
@@ -89,3 +89,9 @@ def test_fock_representation(self):
 
         bs_with_batch = BSgate((0, 1), math.astensor([[1, 2]]), math.astensor([[3], [4], [5]]))
         assert bs_with_batch.fock_array(5, method="stable").shape == (3, 2, 5, 5, 5, 5)
+
+    def test_to_fock_lin_sup(self):
+        dgate = (BSgate((0, 1), 2, 3) + BSgate((0, 1), -2, -3)).to_fock(5)
```
Comment (heltluke) [inline tests/test_lab/test_transformations/test_bsgate.py:94]: Why call this `dgate`?

File: tests/test_training/test_opt_lab_jax.py [hunk 7]
```diff
@@ -263,3 +264,18 @@ def cost_fn(sq):
...
+        # Note: we need to intitialize the cat state with a non-zero value. This is because
+        # the gradients are zero when x is zero.
+        cat_state = Coherent(0, x=0.1, x_trainable=True) + Coherent(0, x=-0.1, x_trainable=True)
+        expected_cat = Coherent(0, x=np.sqrt(np.pi)) + Coherent(0, x=-np.sqrt(np.pi))
+
+        def cost_fn(cat_state):
+            cat_state = cat_state.normalize()
+            return -math.abs(cat_state.fidelity(expected_cat.normalize()))
+
+        opt = OptimizerJax(learning_rate=0.001, stable_threshold=1e-14)
```
Comment (heltluke) [inline tests/test_training/test_opt_lab_jax.py]: How were the `learning_rate` and `stable_threshold` determined? Are they "standard" for MrMustard? If not, is it worth adding a comment about them?

File: mrmustard/math/parameter_set.py
[the remark this answers is not in the document]
Comment (heltluke) [reply mrmustard/math/parameter_set.py:1]: I 100% agree that alphabetization is nice to have.

File: tests/test_lab/test_transformations/test_bsgate.py (hunk 6 above)
Comment (apchytr) [reply tests/test_lab/test_transformations/test_bsgate.py]: Copy paste mistake 😅

**Review by heltluke — APPROVED** [review]
